include/deal.II/grid/tria_info_cache.h

```diff
@@ -0,0 +1,137 @@
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_grid_tria_info_cache_h
+#define dealii_grid_tria_info_cache_h
+
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/point.h>
+
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_info_cache_flags.h>
+#include <deal.II/fe/mapping_q1.h>
+
+#include <cmath>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A class that caches computationally intensive information about a
+ * Triangulation.
+ *
+ * This class attaches a signal to the Triangulation passed at construction
+ * time, and rebuilds all the structures specified in the list of specified
+ * TriangulationInfoCacheFlags, whenever the Triangulation is changed.
+ *
+ * If you try to access one of the objects that has not been cached (due to a
+ * missing TriangulationInfoCacheFlags flag), an Assertion is thrown.
+ *
+ * Notice that this class only notices if the underlying Triangulation has
+ * changed due to a Triangulation::Signals::any_change() signal beeing triggered.
+ *
+ * If the triangulation changes due to a MappingQEulerian being passed to this
+ * class, or because you manually change some vertex locations, then some of
+ * the structures in this class become obsolete, and you will have to call the
+ * method update() manually.
+ *
+ * @author Luca Heltai, 2017.
+ */
+template <int dim, int spacedim>
+class TriangulationInfoCache : public Subscriptor
+{
+public:
+  /**
+   * Constructor.
+   *
+   * You can specify what to cache by passing appropriate
+   * TriangulationInfoCacheFlags. If you provide the optional `mapping`
+   * argument, then this is used whenever a mapping is required.
+   *
+   * @param tria The triangulation to work on
+   * @param flags TriangulationInfoCacheFlags that specify what to cache
+   * @param mapping The mapping to use when computing cached objects
+   */
+  TriangulationInfoCache (const Triangulation<dim,spacedim> &tria,
+                          const TriangulationInfoCacheFlags &flags = cache_nothing,
+                          const Mapping<dim,spacedim> &mapping=StaticMappingQ1<dim,spacedim>::mapping);
+
+  /**
+   * Loop over all TriangulationInfoCacheFlags and rebuilds the specific data
+   * structure.
+   *
+   * The optional parameter can be set to true to update only data that is
+   * dependent on the position of the vertices. This may happen, for example,
+   * if you have a MappingQEulerian class associated to the triangulation. In
```

Comment:
You mean if you have the mapping associated with the *current object*? The triangulation doesn't care.

Reply:
Not unless you are interested in querying about vertex locations. In those cases, you should really ask the mapping and not the triangulation (through `mapping.get_vertex_locations()`)

Reply:
I know. What I'm saying that I think the phrase "if you have a MappingQEulerian class associated to the triangulation" is not quite well defined: we *never associate* mappings with the triangulation. We *use them together*. So the mapping is actually passed to the current object, to be used together with the triangulation, but it is not actually associated with the triangulation.

Reply:
Ok. I adjusted the documentation a bit.